### ISSUE TITLE
Training speed is too slow

### ISSUE THREAD

**Issue by LEECHOONGHO** [facebookresearch/fairscale#959]
Hello, I am conducting multi-GPU training using fairseq's ShardedDDP and OSS.

It is impressive that all gpus consume almost the same memory.
1. However, it takes almost 20 minutes for the training routine to begin.
2. And the training speed seems to be 2-3 times slower than single-gpu training when the batch size/gpu is equal.
3. And The gpu utils seen in nvidia-smi seems to be unstable.

![Honeycam 2022-03-17 12-27-02](https://user-images.githubusercontent.com/44384060/158732749-a25bcc7c-63ff-41e0-b140-3004ef9507b4.gif)

The code I used is like below. Does anyone know if there's something wrong with my usage of function or if there's any way to improve speed?

```
from fairscale.nn.data_parallel import ShardedDataParallel as ShardedDDP
from fairscale.optim.oss import OSS
from torch.cuda.amp import autocast, GradScaler
import torch.distributed as dist
import torch.multiprocessing as mp

mp.spawn(
    train,
    args=(config, train_df, port),
    nprocs=min(torch.cuda.device_count(), config.world_size),
    join=True,
    )

def train(
    rank,
    config,
    train_df,
    port,
):
    
    world_size = config.world_size
    torch.manual_seed(seed)
    torch.cuda.set_device(rank)
    
    
    dist.init_process_group(backend='nccl', init_method=f"tcp://localhost:{str(port)}", rank=rank, world_size=world_size)
    if rank == 0:
        writer = SummaryWriter(log_path)

    train_dataset = DatasetSorted(config, train_df)
    collator = PadCollate()
    model = ASRModel(config).to(rank)

    optimizer = OSS(
        params=model.parameters(),
        optim=torch.optim.Adagrad,
        **config.base_optimizer_arguments
    )
    
    model = ShardedDDP(model, optimizer)

    # Define Sampler & Data Loader
    train_sampler = DistributedSortedDynamicBatchSampler(train_dataset, config, num_replicas=world_size, rank=rank, shuffle=True)

    train_dataloader = torch.utils.data.DataLoader(
        train_dataset,
        shuffle=(train_sampler is None),
        num_workers=config.num_worker, 
        collate_fn=collator,
        pin_memory=False,
        batch_sampler=train_sampler
    )
    
    scaler = GradScaler(enabled=config.fp16_run)
    
    scheduler = Scheduler(config)

    model.train()

    step = 0
    lr = config.init_lr

    for epoch_ in range(config.epochs):
        epoch = epoch_ + 1
        
        lr = scheduler.schedule_by_epoch(optimizer, lr, epoch)

        for inputs in train_dataloader:                                    
            step += 1

            with autocast(enabled=config.fp16_run):
                inputs = {k: v.to(rank) if torch.is_tensor(v) else v for k, v in inputs.items()}

                predictions = model(inputs)
                loss = model.compute_loss(predictions, inputs)

            optimizer.zero_grad()

            scaler.scale(loss).backward()
            scaler.unscale_(optimizer)

            # Clipping gradients to avoid gradient explosion
            grad_norm = clip_grad_value_(model.parameters(), None)

            lr = scheduler.schedule_by_step(optimizer, lr, step)
            
            scaler.step(optimizer)
            scaler.update()

            if rank == 0:
                log_info(config, epoch, step, writer, logger, loss)
                save_model(model, config, epoch, step, optimizer, lr)

    dist.destroy_process_group()
    writer.flush()
    writer.close()

**Comment by min-xu-ai** [comment]
cc: @blefaudeux 

Thanks for the nice screen shot gif and the nice reproduction script. Have you used Nvidia or pytorch profiling tool? Once you use that tool, you can dump some trace file and load it into chrome to see what event takes how long, etc. From that, you can usually find out what's slow.

**Comment by blefaudeux** [comment]
hey there, thanks for the report and the GIF @LEECHOONGHO ! This code has not been maintained for a while I think, you can use FSDP with no reshard to get a similar featureset, but I can comment from memory. 20 minutes to start is not normal, and the GPU use being very low is not normal either, something which could explain is a codebase which checkpoints a lot (looks like you're doing this for every epoch, and I'm not sure how long your epoch is). 

For all sharding trainers, the checkpointing to a single rank is bound to take some extra time (not 20 minutes for a small model though, this may be an issue with your initial process init) because you have to exchange the shards and rebuild a monolithic state. May be what happens here, hard to say for sure but it would be my first guess

**Comment by LEECHOONGHO** [comment]
@blefaudeux  @min-xu-ai  Thank you for your reply.
The time taken for start training was the cost of data loading and sorting.
And the low GPU usage, it maybe caused by Loss function(STC loss)

To improve training speed, I am using FSDP as you suggested.

I wrapped model by FSDP and set mixed_precision True. Not using OSS.
And not using GradScaler, only use autocast for forward progress.

Is there any way to increase the speed or reduce the memory further in  4 GPU training?

```
    model = ASRModel(config).to(rank)

    model = FSDP(
        model, 
        mixed_precision=config.fp16_run,
    )
    
    optimizer = torch.optim.Adam(
        params=model.parameters(),
        **config.base_optimizer_arguments
    )

    try:
        for epoch_ in range(config.epochs):
            epoch = epoch_ + 1

            lr = scheduler.schedule_by_epoch(optimizer, lr, epoch)

            for inputs in train_dataloader:                                    
                step += 1

                with autocast(enabled=config.fp16_run):
                    inputs = {k: v.to(rank) if torch.is_tensor(v) else v for k, v in inputs.items()}

                    predictions = model(inputs)
                    loss = model.compute_loss(predictions, inputs)

                optimizer.zero_grad()
       
                loss['total_loss'].backward()

                # Clipping gradients to avoid gradient explosion
                grad_norm = clip_grad_value_(model.parameters(), None)

                lr = scheduler.schedule_by_step(optimizer, lr, step)      

                optimizer.step()
```

**Comment by blefaudeux** [comment]
- reduce memory: reshard_after_forward = True (slower), or activation checkpointing, or offload
- more speed: reshard_after_forward = False, no clip grad norm (this is actually a little slow. Sidenote: I'm not sure that your implementation is correct when the gradients are sharded, the norm will differ on all agents), faster compute (sometimes depends on the dimensions of your data/model), fused layers (see [xformers](https://github.com/facebookresearch/xformers) for instance)
